### PR TITLE
pick moves using normal distribution

### PR DESCRIPTION
**Pull request by jiteshk23**
Hey @singaurav,  just completed the first cut of pick moves using normal distribution. Pull the changes locally and try running on your machine. Let me know your thoughts.

### CODE REVIEW

**Comment by singaurav** [comment]
@jiteshk23 I have an idea to make this generic. I think we can can generate a pdf(probability for each move) for move vector and write a generic function that takes a pdf and a count, and samples the moves. This way we will have complete runtime control over the sampling. And the RANDOM, NORMAL or any other sampling can be done by just generating the appropriate pdf. The pdf generating functions can also be made configurable. For ex we can pass mean and standard deviation to func which generates normal pdf. What say?

**Comment by jiteshk23** [comment]
Yeah sounds good. I'll try to implement this today.

**Comment by jiteshk23** [comment]
@singaurav I have created 2 functions out of the previous. But still unable to absract out the distribution. I am unable to find the base class for distributions in cpp <random> header. [This](https://stackoverflow.com/questions/32907254/uniform-random-distribution-base-class-for-both-int-and-double) might be of some help. I'll look into it. Will need to brush up on cpp templates I guess.

**Comment by singaurav** [comment]
My suggestion would be to create a sub class of Move called SampleMove which has two extra fields, sample_prob(this is p of move being Sampled) and index(this is the index of the original move vector in GameInfo. Now we write vector<SampleMove> sample ( vector<SampleMove>, count) and then we can write functions like vector<SampleMove> uniform(vector<Move>), vector<SampleMove> normal(vector<Move>, mean, std dev). And the client can simply call one of the latter functions and pass the SampleMove vector to the first function to actually get the sample. Let me know what you think.

**Comment by singaurav** [comment]
First function is
```vector<SampleMove> sample ( vector<SampleMove>, count) ``` and the latter functions are 
```vector<SampleMove> uniform(vector<Move>)``` and 
```vector<SampleMove> normal(vector<Move>, mean, std dev)```

**Comment by jiteshk23** [comment]
Actually I wanted to keep the functions get_distribution and sample_indices generic so that they can be used to sample any vector and not just `vector<Move>`. Would have eventually moved them to a Utilities class.

In terms of number of functions and client usage, it is similar to what you are suggesting
1. sample_indices(count, d)
2. d = get_normal_distribution(count)
3. d = get_uniform_distribution(count)

I think I have a solution now. I'll write a wrapper on top of 2 and 3.
get_distribution(name) // name could be 'UNIFORM', 'NORMAL' etc.

And then 'sample_moves' can specify a 'distribution_name' paramter.
```
sample_moves(......., distribution_name)
{
    .....
    distribution = get_distribution(distribution_name, count);
    .....
}
```

File: src/gameinfo.cpp [hunk 1]
```diff
@@ -162,94 +164,18 @@ unsigned GameWithSampledMoves::sample_count<PERC, double>(double perc) {
 
 template <>
 void GameWithSampledMoves::sample_moves<RANDOM>(unsigned count) {
```
Comment (jiteshk23) [inline src/gameinfo.cpp:166]: @singaurav may be we should change RANDOM to UNIFORM

Comment (singaurav) [reply src/gameinfo.cpp:166]: Agreed, uniform makes more sense.

**Review by singaurav — APPROVED** [review]

**Pull request merged**